Source/DafnyLanguageServer/Handlers/DafnyHoverHandler.cs

```diff
@@ -24,30 +30,145 @@ public class DafnyHoverHandler : HoverHandlerBase {
       };
     }
 
-    public async override Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
+    public override async Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
       logger.LogTrace("received hover request for {Document}", request.TextDocument);
       var document = await documents.GetDocumentAsync(request.TextDocument);
       if (document == null) {
         logger.LogWarning("the document {Document} is not loaded", request.TextDocument);
         return null;
       }
+      var diagnosticHoverContent = GetDiagnosticsHover(document, request.Position);
       if (!document.SymbolTable.TryGetSymbolAt(request.Position, out var symbol)) {
         logger.LogDebug("no symbol was found at {Position} in {Document}", request.Position, request.TextDocument);
+      }
+
+      var symbolHoverContent = symbol != null ? CreateSymbolMarkdown(symbol, cancellationToken) : null;
+      if (diagnosticHoverContent == null && symbolHoverContent == null) {
         return null;
       }
-      return CreateHover(symbol, cancellationToken);
+
+      var hoverContent = diagnosticHoverContent == null ? "" : diagnosticHoverContent;
+      hoverContent += symbolHoverContent == null ? (hoverContent != "" ? "  \n" : "") : symbolHoverContent;
+      return CreateMarkdownHover(hoverContent);
+    }
+
+    private string? GetDiagnosticsHover(DafnyDocument document, Position position) {
+      foreach (var node in document.VerificationNodeDiagnostic.Children.OfType<TopLevelDeclMemberNodeDiagnostic>()) {
+        if (node.Range.Contains(position)) {
+          var assertionBatchCount = node.AssertionBatchCount;
+          var assertionBatchIndex = -1;
+          var information = "";
```

Comment:
There can be multiple assertions on the same token, from different assertion batches. This field records them all.